6397 rebalance test matrix

ttoomey opened the pull request:
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

[GH Issue](https://github.com/open-path/Green-River/issues/6397)

* should get our CI runs down to around 45 min.
* requires occasional manual profiling and re bucketing
* note, this requires adding some tags to our specs that will be managed automatically. Not every spec is tagged, only the slow ones (> 10 secs). Faster tests are all run in a default bucket together
* This surfaced some test failures which may be pre-existing issues in the tests related to missing HMIS forms. I think we see them here because the test order has changed. 

Rebalancing procedure:
1) run rspec with profiling enabled, output results to json (maybe this should happen on every run?)
2) download profiles from GH artifacts into rspec_profiles directory
3) run `rails ci:build_bucket_assignments[/tmp/rspec_profiles]` which updates `.github/rspec_buckets.json`
4) if the rake task added new buckets, add them to the github text matrix


## Type of change
[//]: # 'remove options that are not relevant'
- [x] Code clean-up / housekeeping

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [ ] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ttoomey committed:
5755-run-e2e-tests-in-ci

ttoomey committed:
first try ci e2e

ttoomey committed:
try 2

ttoomey committed:
take 3

ttoomey committed:
try 4

ttoomey committed:
take 5

ttoomey committed:
take 6

ttoomey committed:
try 7

ttoomey committed:
take 8

ttoomey committed:
take 9

ttoomey committed:
take 10

ttoomey committed:
add back real ci

ttoomey committed:
integrate system tests back into existing config

ttoomey committed:
syntax

ttoomey committed:
syntax

ttoomey committed:
wip

ttoomey committed:
enter the matrix

ttoomey committed:
fix

ttoomey committed:
fixes

ttoomey committed:
add caching

ttoomey committed:
fix

ttoomey committed:
fix

ttoomey committed:
fixes

ttoomey committed:
take 11

ttoomey committed:
fix

ttoomey committed:
take 12

ttoomey committed:
take 13

ttoomey committed:
take 14

ttoomey committed:
rollback

ttoomey committed:
5755-run-e2e-tests-in-ci

ttoomey committed:
fix

ttoomey committed:
5755-run-e2e-tests-in-ci

ttoomey committed:
wip

ttoomey committed:
wip

ttoomey committed:
fix

ttoomey committed:
Merge branch 'release-125' into 5755-run-e2e-tests-in-ci

ttoomey committed:
fixes

ttoomey committed:
5755-run-e2e-tests-in-ci

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
add tags

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalence-test-matrix

ttoomey committed:
6397-rebalance-test-matrix

ttoomey commented on drivers/hmis/spec/requests/hmis/form_definition_rules_spec.rb:
```diff
@@ -11,6 +11,10 @@
   include_context 'hmis base setup'
 
+  before(:all) do
+    ::HmisUtil::JsonForms.seed_all
```
this started failing, seems like an order-dependent issue

eanders replied:
Does it make sense to move this into `hmis base setup`?

ttoomey replied:
Yes, I think so. I'll give it a try and and see what breaks

ttoomey replied:
hmm that looks like a bigger change actually since there's like 90 tests that use that. I might hold off on that for now

eanders commented on lib/tasks/ci.rake:
```diff
@@ -0,0 +1,97 @@
+      # Skip specs that run for less than 10 seconds
+      next if total_time < 10
+
+      if current_bucket[:total_time] + total_time > (60 * 40)
```
is `60 * 40` 40 minutes?  Might be worth adding a comment or some variables to indicate what this is checking.

eanders reviewed (commented):
@ttoomey I _think_ this looks like a good approach.  
In some ways I'd prefer to drop the lists of tests per bucket into some sort of text or yaml file so we don't need to make code changes to tests and then guess/check/adjust those as new tests are added, but what you've written seems pretty light weight for now.

Happy to give this a whirl and revisit in a month or two.

ttoomey committed:
Merge branch 'release-125' into 6397-rebalance-test-matrix

ttoomey committed:
Revert "add tags"

This reverts commit e97c7c71982c9a4933cfcdf3b869a95f05678e3f.

ttoomey committed:
bin packing

ttoomey committed:
6397-rebalance-test-matrix

ttoomey committed:
6397-rebalance-test-matrix

ttoomey committed:
6397-rebalance-test-matrix

fix s3 uploader failure

ttoomey committed:
6397-rebalance-test-matrix

test fix

ttoomey commented:
> @ttoomey I _think_ this looks like a good approach. In some ways I'd prefer to drop the lists of tests per bucket into some sort of text or yaml file so we don't need to make code changes to tests and then guess/check/adjust those as new tests are added, but what you've written seems pretty light weight for now.

Thanks @eanders - I changed the approach so that we don't need to commit the tags. Now it keeps the groups in a json file (.github/rspec_buckets.json) and modifies the tests in the CI container right before running them. Still plenty of room for improvement though

I also adjusted the bucket approach to try and get the runtime down closer to 30 minutes.

ttoomey commented on app/models/aws_s3.rb:
```diff
@@ -148,11 +148,11 @@ def put(file_name:, prefix:)
+    args.merge!(content_type: content_type) if content_type
+    # TODO: document why we skip encryption here
+    args.merge!(server_side_encryption: 'AES256') unless Rails.env.development? || Rails.env.test?
+    obj.put(**args)
```
@eanders 
This addresses a failure in `rspec ./spec/models/aws_s3_spec.rb:48` which was added [in this pr](https://github.com/greenriver/hmis-warehouse/pull/4553/files#diff-ded50fd5f0dc4498ae5af0ac5b785601d5319da6b6be9862628d235917ed1a3eR154)

ttoomey replied:
[extracted this fix to pr](https://github.com/greenriver/hmis-warehouse/pull/4572)

ttoomey committed:
6397-rebalance-test-matrix

eanders commented on app/models/aws_s3.rb:
```diff
@@ -148,11 +148,11 @@ def put(file_name:, prefix:)
-    end
+    args = { body: content }
+    args.merge!(content_type: content_type) if content_type
+    # TODO: document why we skip encryption here
```
So, I believe we're skipping server side encryption for test and development because minio didn't (or doesn't) support it, or it was fairly complex to setup and didn't seem necessary for local development or testing.  Might be better to inspect the hostname, and ensure it's enabled for real s3 urls.

ttoomey committed:
6397-rebalance-test-matrix

ttoomey committed:
extract aws test bugfix from 6397-rebalance-test-matrix

ttoomey committed:
Merge branch 'release-125' into 6397-rebalance-test-matrix

ttoomey committed:
Merge branch 'fix-aws-test-bug' into 6397-rebalance-test-matrix

eanders reviewed (approved):
Pending verification of test counts

ttoomey committed:
Merge branch 'release-125' into 6397-rebalance-test-matrix

ttoomey commented:
I compared between the two branches. The total time is very similar but it looks like there's a small discrepancy, there are 4 more examples counted in release-125 than in this branch for some reason. @eanders - do you think I should try and track that down or is it close enough?

I could re-run profiling on r125 and this branch and diff the specs to track it down, might take an hour or so. Note, the last time I ran profiling, we had 5,148 total examples.

### This branch with new buckets
| bucket  | minutes | examples |
|---------|---------|-------|
| 1       | 30      | 903   |
| 2       | 29      | 691   |
| 3       | 26      | 131   |
| 4       | 30      | 906   |
| default | 20      | 2523  |
| system  | 3       | 10    |
| okta    | 1       | 21    |
| **Total** | **139**  | **5185** |

### release-125
| bucket | minutes | examples |
|--------|---------|-------|
| a-g    | 48      | 1265  |
| h-i    | 68      | 2512  |
| j-z    | 2       | 19    |
| spec   | 12      | 1362  |
| system | 3       | 10    |
| okta   | 0       | 21    |
| **Total** | **133**  | **5189** |

eanders committed:
Merge branch 'release-125' into 6397-rebalance-test-matrix

ttoomey committed:
Merge branch '6397-rebalance-test-matrix' of github.com:greenriver/hmis-warehouse into 6397-rebalance-test-matrix

ttoomey committed:
re-profile

ttoomey committed:
revert profile

ttoomey commented:
the four tests were from this spec file
drivers/hmis/lib/spec/fix_enrollment_dates_spec.rb

It's outside of the expected directory pattern for specs so it's excluded. I think that's okay, it looks like it was a test for a one-time data fix
